parser.d.ts

```diff
@@ -0,0 +1,136 @@
+export type AstNodeType =
+  | "alternative"
+  | "anchor"
+  | "characterClass"
+  | "characterClassEscape"
+  | "characterClassRange"
+  | "disjunction"
+  | "dot"
+  | "group"
+  | "quantifier"
+  | "reference"
+  | "unicodePropertyEscape"
+  | "value";
+
+export type Base<T extends AstNodeType> = {
+  range: [number, number];
+  raw: string;
+  type: T;
+};
+
+export type AstNode =
+  | Alternative
+  | Anchor
+  | CharacterClass
+  | CharacterClassEscape
+  | CharacterClassRange
+  | Disjunction
+  | Dot
+  | Group
+  | Quantifier
+  | Reference
+  | UnicodePropertyEscape
```

Comment:
The unicode related types are only available when the `unicodePropertyEscape` flag is provided. Is there a way to encode this in the type system?

Reply:
I think I thought about this and it _might_ be possible with generics, but it would make the types a lot more complicated. Will have another think

Reply:
In Babel we found that parametrizing the AST types based on the parser options is possible, but it leads to slower TypeScript checks (however, we have a way larger number of possible AST nodes).

This is an example of what I tried to do, with some nodes that are only valid when the `ESTree` option was either `true` or `false`: https://gist.github.com/nicolo-ribaudo/409cc9c406531e20abd29d291e9693db.

Reply:
thanks. will have a go at this. A slight performance hit at build time is worth it for the better type safety IMO

Reply:
I tried it and it looks like it works well. I think we can do something similar with the `namedGroups` feature and make

```ts
export type Reference = Base<"reference"> &
  (
    | {
        matchIndex: number;
        name?: undefined;
      }
    | {
        matchIndex?: undefined;
        name: string;
      }
  );
```
better based on that. Will look at that next

Reply:
done